include/rcutils/types/hash_map.h

```diff
@@ -0,0 +1,422 @@
+ * \return `RCUTILS_RET_ERROR` if an unknown error occurs
+ */
+RCUTILS_PUBLIC
+RCUTILS_WARN_UNUSED
+rcutils_ret_t
+rcutils_hash_map_unset(rcutils_hash_map_t * hash_map, const void * key);
+
+/// Get whether or not a key exists.
+/**
+ * Returns true if the provided key exists in the hash_map or false if it does not or
+ * if the hash_map or key are invalid.
+ * In all cases no error message is set.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] hash_map rcutils_hash_map_t to be searched
+ * \param[in] key hash_map key, must be null terminated c string
+ * \return `true` if key is in the hash_map, or
+ * \return `false` if key is not in the hash_map, or
+ * \return `false` for invalid arguments, or
+ * \return `false` if the hash_map is invalid
+ */
+RCUTILS_PUBLIC
+bool
+rcutils_hash_map_key_exists(const rcutils_hash_map_t * hash_map, const void * key);
+
+/// Get value given a key.
+/**
+ * This function can be used to retrieve a shallow copy of the stored data. The data
+ * pointer must point to a section of memory large enough to copy the full size of
+ * the data being stored, which is specified when the hash_map in initialized.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] hash_map rcutils_hash_map_t to be searched
+ * \param[in] key hash_map key to look up the data for
+ * \param[out] data A copy of the data stored in the map
+ * \return `RCUTILS_RET_OK` if successful, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENT` for invalid arguments, or
+ * \return `RCUTILS_RET_NOT_INITIALIZED` if the hash_map is invalid, or
+ * \return `RCUTILS_RET_NOT_FOUND` if the key doesn't exist in the map, or
+ * \return `RCUTILS_RET_ERROR` if an unknown error occurs
+ */
+RCUTILS_PUBLIC
+rcutils_ret_t
+rcutils_hash_map_get(const rcutils_hash_map_t * hash_map, const void * key, void * data);
+
+/// Get the next key in the hash_map, unless NULL is given, then get the first key.
+/**
+ * This function allows you to iteratively get each key/value pair in the hash_map.
+ *
+ * If NULL is given for the previous_key, then the first key in the hash_map is returned.
+ * If that returned key if given as the previous_key for the next call to this function,
+ * then the next key in the hash_map is returned.
+ * If there are no more keys in the hash_map or if the given key is not in the hash_map,
+ * an error will be returned.
+ *
+ * The order of the keys in the hash_map is arbitrary and if the hash_map is modified
+ * between calls to this function the behavior is undefined.
+ * If the hash_map is modifeid then iteration should begin again by passing NULL to
+ * get the first key again.
```

Comment:
@nburek hmm, isn't this statement against [your use case](https://github.com/ros2/rcl/pull/350/files#diff-bf8aead5abb376e0ca7e38051163fb9dR120) in rcl?

Reply:
Yes, I'll correct the rcl review.